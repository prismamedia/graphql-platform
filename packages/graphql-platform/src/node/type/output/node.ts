import type { Name } from '@prismamedia/graphql-platform-utils';
import * as utils from '@prismamedia/graphql-platform-utils';
import { MGetter, MMethod } from '@prismamedia/memoize';
import * as graphql from 'graphql';
import assert from 'node:assert';
import type { Except } from 'type-fest';
import type { BrokerInterface } from '../../../broker-interface.js';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { Node } from '../../../node.js';
import { Leaf, UniqueReverseEdge, type Component } from '../../definition.js';
import type { OperationContext } from '../../operation/context.js';
import {
  NodeSelection,
  mergeSelectionExpressions,
  type NodeSelectedSource,
  type NodeSelectedValue,
  type SelectionExpression,
} from '../../statement/selection.js';
import {
  EdgeHeadOutputType,
  LeafOutputType,
  MultipleReverseEdgeCountOutputType,
  MultipleReverseEdgeHeadOutputType,
  UniqueReverseEdgeHeadOutputType,
  VirtualOutputType,
  isMultipleReverseEdgeOutputType,
  type ComponentOutputType,
  type MultipleReverseEdgeOutputType,
  type NodeFieldOutputType,
  type ReverseEdgeOutputType,
  type ThunkableVirtualOutputConfig,
  type ThunkableVirtualOutputConfigsByName,
} from './node/field.js';

export * from './node/field.js';

export type GraphQLSelectionAST =
  | graphql.DocumentNode
  | graphql.FragmentDefinitionNode
  | graphql.InlineFragmentNode
  | graphql.SelectionSetNode;

export type GraphQLSelectionContext = Partial<
  Pick<graphql.GraphQLResolveInfo, 'fragments' | 'variableValues'>
>;

/**
 * A stringified fragment under one of these 4 forms:
 * * { _id id category { id } }
 * * ... { _id id category { id } }
 * * ... on Article { _id id category { id } }
 * * fragment MyFragment on Article { _id id category { id } }
 */
export type GraphQLFragment = string;

export type PartialGraphQLResolveInfo = Pick<
  graphql.GraphQLResolveInfo,
  'fieldNodes' | 'returnType' | 'path' | 'fragments' | 'variableValues'
>;

export const isPartialGraphQLResolveInfo = (
  maybePartialGraphQLResolveInfo: unknown,
): maybePartialGraphQLResolveInfo is PartialGraphQLResolveInfo =>
  utils.isPlainObject(maybePartialGraphQLResolveInfo) &&
  Array.isArray(maybePartialGraphQLResolveInfo['fieldNodes']) &&
  maybePartialGraphQLResolveInfo['fieldNodes'].length > 0 &&
  graphql.isOutputType(maybePartialGraphQLResolveInfo['returnType']) &&
  utils.isPath(maybePartialGraphQLResolveInfo['path']);

export type RawNodeSelection<
  TSource extends NodeSelectedSource = any,
  TValue extends NodeSelectedValue = TSource,
> =
  | NodeSelection<TSource, TValue>
  | ReadonlyArray<Component | Component['name']>
  | GraphQLFragment
  | GraphQLSelectionAST
  | PartialGraphQLResolveInfo;

export interface NodeOutputTypeConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TBroker extends BrokerInterface,
  TContainer extends object,
> {
  /**
   * Optional, add some "virtual" fields whose value is computed from the components' value
   *
   * They are called "virtual" because they are not persisted
   */
  virtualFields?: ThunkableVirtualOutputConfigsByName<
    TRequestContext,
    TConnector,
    TBroker,
    TContainer
  >;

  /**
   * Optional, fine-tune the GraphQL output type
   */
  graphql?: Except<
    graphql.GraphQLObjectTypeConfig<
      Readonly<NodeSelectedValue>,
      TRequestContext
    >,
    'name' | 'description' | 'fields'
  >;
}

export class NodeOutputType {
  public readonly name: Name;

  readonly #config?: NodeOutputTypeConfig<any, any, any, any>;
  readonly #configPath: utils.Path;

  public constructor(public readonly node: Node) {
    // config
    {
      this.#config = node.config.output || undefined;
      this.#configPath = utils.addPath(node.configPath, 'output');

      utils.assertNillablePlainObject(this.#config, this.#configPath);
    }

    this.name = node.name;
  }

  public toString(): string {
    return this.name;
  }

  @MGetter
  public get componentFieldsByName(): ReadonlyMap<
    ComponentOutputType['name'],
    ComponentOutputType
  > {
    return new Map(
      Array.from(this.node.componentSet, (component) => {
        const field =
          component instanceof Leaf
            ? new LeafOutputType(this, component)
            : new EdgeHeadOutputType(this, component);

        return [field.name, field];
      }),
    );
  }

  @MGetter
  public get leafFieldsByName(): ReadonlyMap<
    LeafOutputType['name'],
    LeafOutputType
  > {
    return new Map(
      this.componentFieldsByName
        .entries()
        .filter(
          (entry): entry is [string, LeafOutputType] =>
            entry[1] instanceof LeafOutputType,
        ),
    );
  }

  @MGetter
  public get edgeFieldsByName(): ReadonlyMap<
    EdgeHeadOutputType['name'],
    EdgeHeadOutputType
  > {
    return new Map(
      this.componentFieldsByName
        .entries()
        .filter(
          (entry): entry is [string, EdgeHeadOutputType] =>
            entry[1] instanceof EdgeHeadOutputType,
        ),
    );
  }

  @MGetter
  public get reverseEdgeFieldsByName(): ReadonlyMap<
    ReverseEdgeOutputType['name'],
    ReverseEdgeOutputType
  > {
    return new Map(
      this.node.reverseEdgeSet.values().flatMap((reverseEdge) => {
        const fields: ReverseEdgeOutputType[] = [];

        if (reverseEdge instanceof UniqueReverseEdge) {
          fields.push(new UniqueReverseEdgeHeadOutputType(this, reverseEdge));
        } else {
          fields.push(
            new MultipleReverseEdgeHeadOutputType(this, reverseEdge),
            new MultipleReverseEdgeCountOutputType(this, reverseEdge),
          );
        }

        return fields.map((field) => [field.name, field]);
      }),
    );
  }

  @MGetter
  public get uniqueReverseEdgeFieldsByName(): ReadonlyMap<
    UniqueReverseEdgeHeadOutputType['name'],
    UniqueReverseEdgeHeadOutputType
  > {
    return new Map(
      this.reverseEdgeFieldsByName
        .entries()
        .filter(
          (entry): entry is [string, UniqueReverseEdgeHeadOutputType] =>
            entry[1] instanceof UniqueReverseEdgeHeadOutputType,
        ),
    );
  }

  @MGetter
  public get multipleReverseEdgeFieldsByName(): ReadonlyMap<
    MultipleReverseEdgeOutputType['name'],
    MultipleReverseEdgeOutputType
  > {
    return new Map(
      this.reverseEdgeFieldsByName
        .entries()
        .filter((entry): entry is [string, MultipleReverseEdgeOutputType] =>
          isMultipleReverseEdgeOutputType(entry[1]),
        ),
    );
  }

  @MGetter
  public get virtualFieldsByName(): ReadonlyMap<
    VirtualOutputType['name'],
    VirtualOutputType
  > {
    const fields: VirtualOutputType[] = [];

    const currentFieldNameSet = new Set([
      ...this.componentFieldsByName.keys(),
      ...this.reverseEdgeFieldsByName.keys(),
    ]);

    // virtual-fields
    this.node.features.forEach(({ config, configPath }) => {
      const outputConfig = config.output;
      const outputConfigPath = utils.addPath(configPath, 'output');

      utils.assertNillablePlainObject(outputConfig, outputConfigPath);

      if (!outputConfig?.virtualFields) {
        return;
      }

      const virtualFieldConfigsByName = utils.resolveThunkable(
        outputConfig.virtualFields,
        this.node,
      );
      const virtualFieldConfigsByNamePath = utils.addPath(
        outputConfigPath,
        'virtualFields',
      );

      utils.assertPlainObject(
        virtualFieldConfigsByName,
        virtualFieldConfigsByNamePath,
      );

      utils.aggregateGraphError<
        [utils.Name, ThunkableVirtualOutputConfig],
        void
      >(
        Object.entries(virtualFieldConfigsByName),
        (_, [virtualFieldName, thunkableVirtualFieldConfig]) => {
          const virtualFieldConfig = thunkableVirtualFieldConfig;
          const virtualFieldConfigPath = utils.addPath(
            virtualFieldConfigsByNamePath,
            virtualFieldName,
          );

          const virtualField = new VirtualOutputType(
            this,
            virtualFieldName,
            virtualFieldConfig,
            virtualFieldConfigPath,
          );

          if (currentFieldNameSet.has(virtualField.name)) {
            throw new utils.GraphError(
              `At least 1 field already have this name`,
              { path: virtualFieldConfigPath },
            );
          } else {
            currentFieldNameSet.add(virtualField.name);
          }

          fields.push(virtualField);
        },
        undefined,
        { path: virtualFieldConfigsByNamePath },
      );
    });

    return new Map(fields.map((field) => [field.name, field]));
  }

  @MGetter
  public get fieldsByName(): ReadonlyMap<
    NodeFieldOutputType['name'],
    NodeFieldOutputType
  > {
    return new Map<NodeFieldOutputType['name'], NodeFieldOutputType>([
      ...this.componentFieldsByName,
      ...this.reverseEdgeFieldsByName,
      ...this.virtualFieldsByName,
    ]);
  }

  @MMethod()
  public getGraphQLObjectType(): graphql.GraphQLObjectType {
    assert(this.node.isPublic(), `The "${this.node}" node is private`);

    return new graphql.GraphQLObjectType({
      ...this.#config?.graphql,
      name: this.name,
      description: this.node.description,
      fields: () =>
        Object.fromEntries(
          utils.aggregateGraphError<
            NodeFieldOutputType,
            [string, graphql.GraphQLFieldConfig<any, any>][]
          >(
            this.fieldsByName.values(),
            (entries, field) =>
              field.isPublic()
                ? [...entries, [field.name, field.getGraphQLFieldConfig()]]
                : entries,
            [],
            { path: this.#configPath },
          ),
        ),
    });
  }

  @MMethod()
  public validate(): void {
    utils.aggregateGraphError<NodeFieldOutputType, void>(
      this.fieldsByName.values(),
      (_, field) => field.validate(),
      undefined,
      { path: this.#configPath },
    );

    this.node.isPublic() && this.getGraphQLObjectType();
  }

  public getFieldByName(name: string, path?: utils.Path): NodeFieldOutputType {
    const field = this.fieldsByName.get(name);
    if (!field) {
      throw new utils.UnexpectedValueError(
        `${this.node.indefinite}'s field among "${[
          ...this.fieldsByName.keys(),
        ].join(', ')}"`,
        name,
        { path },
      );
    }

    return field;
  }

  public selectComponents(
    componentOrNames: ReadonlyArray<Component | Component['name']>,
    _operationContext?: OperationContext,
    path?: utils.Path,
  ): NodeSelection {
    if (!Array.isArray(componentOrNames)) {
      throw new utils.UnexpectedValueError(
        `an array of component`,
        componentOrNames,
        { path },
      );
    }

    return new NodeSelection(
      this.node,
      mergeSelectionExpressions(
        componentOrNames.map(
          (componentOrName, index) =>
            this.node.ensureComponent(
              componentOrName,
              utils.addPath(path, index),
            ).selection,
        ),
        path,
      ),
    );
  }

  public selectGraphQLFragment(
    fragment: GraphQLFragment,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    if (typeof fragment !== 'string') {
      throw new utils.UnexpectedValueError(`a GraphQL fragment`, fragment, {
        path,
      });
    }

    const trimmedFragment = fragment.trim();

    let document: graphql.DocumentNode;
    try {
      document = graphql.parse(
        trimmedFragment.startsWith('fragment ')
          ? // A regular fragment
            trimmedFragment
          : // An anonymous or inline fragment
            trimmedFragment.replace(
              /(^|\.\.\.[^\{]*){/,
              `fragment MyFragment on ${this.node} {`,
            ),
      );
    } catch (cause) {
      throw new utils.UnexpectedValueError(
        `a valid GraphQL fragment`,
        fragment,
        { cause, path },
      );
    }

    return this.selectGraphQLDocumentNode(
      document,
      operationContext,
      selectionContext,
      path,
    );
  }

  @MGetter
  protected get typeNames(): ReadonlySet<graphql.GraphQLObjectType['name']> {
    return new Set([this.name, this.node.deletionOutputType.name]);
  }

  public selectGraphQLDocumentNode(
    ast: graphql.DocumentNode,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    utils.assertGraphQLASTNode(ast, graphql.Kind.DOCUMENT, path);

    const fragmentDefinitions = ast.definitions.filter(
      (definitionNode): definitionNode is graphql.FragmentDefinitionNode =>
        definitionNode.kind === graphql.Kind.FRAGMENT_DEFINITION,
    );

    const fragmentDefinition = fragmentDefinitions.find((definitionNode) =>
      this.typeNames.has(definitionNode.typeCondition.name.value),
    );

    if (!fragmentDefinition) {
      throw new utils.UnexpectedValueError(
        `a GraphQL ${graphql.Kind.FRAGMENT_DEFINITION} on "${this.node}"`,
        ast,
        { path },
      );
    }

    return this.selectGraphQLFragmentDefinitionNode(
      fragmentDefinition,
      operationContext,
      {
        ...selectionContext,
        fragments: {
          ...selectionContext?.fragments,
          ...Object.fromEntries(
            fragmentDefinitions.map((definition) => [
              definition.name.value,
              definition,
            ]),
          ),
        },
      },
      path,
    );
  }

  public selectGraphQLFragmentDefinitionNode(
    ast: graphql.FragmentDefinitionNode,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    utils.assertGraphQLASTNode(ast, graphql.Kind.FRAGMENT_DEFINITION, path);

    if (!this.typeNames.has(ast.typeCondition.name.value)) {
      throw new utils.UnexpectedValueError(
        `a GraphQL ${graphql.Kind.FRAGMENT_DEFINITION} on "${this}"`,
        ast,
        { path },
      );
    }

    return this.selectGraphQLSelectionSetNode(
      ast.selectionSet,
      operationContext,
      selectionContext,
      path,
    );
  }

  public selectGraphQLInlineFragmentNode(
    ast: graphql.InlineFragmentNode,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    utils.assertGraphQLASTNode(ast, graphql.Kind.INLINE_FRAGMENT, path);

    if (
      ast.typeCondition &&
      !this.typeNames.has(ast.typeCondition.name.value)
    ) {
      throw new utils.UnexpectedValueError(
        `a GraphQL ${graphql.Kind.INLINE_FRAGMENT} on "${this}"`,
        ast,
        { path },
      );
    }

    return this.selectGraphQLSelectionSetNode(
      ast.selectionSet,
      operationContext,
      selectionContext,
      path,
    );
  }

  public selectGraphQLSelectionSetNode(
    ast: graphql.SelectionSetNode,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    utils.assertGraphQLASTNode(ast, graphql.Kind.SELECTION_SET, path);

    return new NodeSelection(
      this.node,
      mergeSelectionExpressions(
        ast.selections.flatMap<SelectionExpression>((ast) => {
          /**
           * "[...] the field or fragment must not be queried if either the @skip condition is true or the @include condition is false [...]"
           *
           * @see http://spec.graphql.org/draft/#note-f3059
           */
          if (
            ast.directives?.some((directive) =>
              directive.name.value === 'skip'
                ? directive.arguments?.find(
                    (argument) =>
                      argument.name.value === 'if' &&
                      graphql.valueFromASTUntyped(
                        argument.value,
                        selectionContext?.variableValues,
                      ) === true,
                  )
                : directive.name.value === 'include'
                  ? directive.arguments?.find(
                      (argument) =>
                        argument.name.value === 'if' &&
                        graphql.valueFromASTUntyped(
                          argument.value,
                          selectionContext?.variableValues,
                        ) === false,
                    )
                  : false,
            )
          ) {
            return [];
          }

          switch (ast.kind) {
            case graphql.Kind.FIELD: {
              const fieldAlias = ast.alias?.value;
              const fieldName = ast.name.value;
              const fieldKey = fieldAlias ?? fieldName;

              // Handle the "__typename" meta field
              if (fieldName === graphql.TypeNameMetaFieldDef.name) {
                return [];
              }

              return this.getFieldByName(
                fieldName,
                path,
              ).selectGraphQLFieldNode(
                ast,
                operationContext,
                selectionContext,
                utils.addPath(path, fieldKey),
              );
            }

            case graphql.Kind.FRAGMENT_SPREAD: {
              const fragmentName = ast.name.value;
              const fragmentDefinition =
                selectionContext?.fragments?.[fragmentName];

              if (!fragmentDefinition) {
                throw new utils.UnexpectedValueError(
                  `the GraphQL fragment definition named "${fragmentName}"`,
                  selectionContext?.fragments,
                  { path },
                );
              }

              return this.selectGraphQLFragmentDefinitionNode(
                fragmentDefinition,
                operationContext,
                selectionContext,
                path,
              ).expressions;
            }

            case graphql.Kind.INLINE_FRAGMENT: {
              return this.selectGraphQLInlineFragmentNode(
                ast,
                operationContext,
                selectionContext,
                path,
              ).expressions;
            }

            default:
              throw new utils.UnreachableValueError(ast, { path });
          }
        }),
        path,
      ),
    );
  }

  public selectGraphQLResolveInfo(
    {
      returnType,
      fieldNodes,
      path,
      fragments,
      variableValues,
    }: PartialGraphQLResolveInfo,
    operationContext?: OperationContext,
  ): NodeSelection {
    const namedReturnType = graphql.getNamedType(returnType);
    if (namedReturnType !== this.getGraphQLObjectType()) {
      throw new utils.UnexpectedValueError(
        `a resolver returning ${this.node.indefinite}`,
        String(namedReturnType),
        { path },
      );
    }

    return this.selectGraphQLSelectionSetNode(
      fieldNodes[0].selectionSet!,
      operationContext,
      { fragments, variableValues },
      path,
    );
  }

  public select(
    rawSelection: RawNodeSelection,
    operationContext?: OperationContext,
    selectionContext?: GraphQLSelectionContext,
    path?: utils.Path,
  ): NodeSelection {
    if (rawSelection == null) {
      throw new utils.UnexpectedValueError(
        `${this.node.indefinite}'s selection`,
        rawSelection,
        { path },
      );
    }

    if (rawSelection instanceof NodeSelection) {
      return rawSelection;
    }

    // Components
    if (Array.isArray(rawSelection)) {
      return this.selectComponents(rawSelection, operationContext, path);
    }

    // GraphQL fragment
    if (typeof rawSelection === 'string') {
      return this.selectGraphQLFragment(
        rawSelection,
        operationContext,
        selectionContext,
        path,
      );
    }

    // GraphQL AST
    if (utils.isPlainObject(rawSelection) && 'kind' in rawSelection) {
      switch (rawSelection.kind) {
        case graphql.Kind.DOCUMENT:
          return this.selectGraphQLDocumentNode(
            rawSelection,
            operationContext,
            selectionContext,
            path,
          );

        case graphql.Kind.FRAGMENT_DEFINITION:
          return this.selectGraphQLFragmentDefinitionNode(
            rawSelection,
            operationContext,
            selectionContext,
            path,
          );

        case graphql.Kind.INLINE_FRAGMENT:
          return this.selectGraphQLInlineFragmentNode(
            rawSelection,
            operationContext,
            selectionContext,
            path,
          );

        case graphql.Kind.SELECTION_SET:
          return this.selectGraphQLSelectionSetNode(
            rawSelection,
            operationContext,
            selectionContext,
            path,
          );

        default:
          throw new utils.UnreachableValueError(rawSelection, { path });
      }
    }

    // GraphQL resolve info
    if (isPartialGraphQLResolveInfo(rawSelection)) {
      return this.selectGraphQLResolveInfo(rawSelection, operationContext);
    }

    throw new utils.UnexpectedValueError(
      `a supported "${this.node}"'s selection`,
      rawSelection,
      { path },
    );
  }

  public selectShape(
    shape: unknown,
    operationContext?: OperationContext,
    path?: utils.Path,
  ): NodeSelection {
    utils.assertPlainObject(shape, path);

    return new NodeSelection(
      this.node,
      mergeSelectionExpressions(
        Object.entries(shape).flatMap<SelectionExpression>(
          ([fieldName, fieldValue]) =>
            this.getFieldByName(fieldName, path).selectShape(
              fieldValue,
              operationContext,
              utils.addPath(path, fieldName),
            ),
        ),
        path,
      ),
    );
  }
}
