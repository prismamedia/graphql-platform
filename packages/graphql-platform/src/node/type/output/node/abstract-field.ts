import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import * as graphql from 'graphql';
import assert from 'node:assert/strict';
import { argsPathKey } from '../../../abstract-operation.js';
import type { OperationContext } from '../../../operation/context.js';
import type {
  NodeSelectedValue,
  SelectionExpression,
} from '../../../statement/selection.js';
import type { GraphQLSelectionContext } from '../node.js';

export interface AbstractNodeFieldOutputTypeConfig {
  name: utils.Name;
  public: boolean;
  description?: utils.OptionalDescription;
  deprecated?: utils.OptionalDeprecation;
}

export abstract class AbstractNodeFieldOutputType<
  TArgs extends utils.Nillable<utils.PlainObject>,
> {
  public abstract readonly name: utils.Name;
  public abstract readonly description?: string;
  public abstract readonly deprecationReason?: string;
  public abstract readonly arguments?: ReadonlyArray<utils.Input>;
  public abstract readonly type: graphql.GraphQLOutputType;

  public abstract isPublic(): boolean;

  @Memoize()
  public getGraphQLFieldConfig(): graphql.GraphQLFieldConfig<
    NodeSelectedValue,
    any,
    any
  > {
    assert(this.isPublic(), `The "${this}" field is private`);

    return {
      ...(this.description && { description: this.description }),
      ...(this.deprecationReason && {
        deprecationReason: this.deprecationReason,
      }),
      ...(this.arguments?.length && {
        args: utils.getGraphQLFieldConfigArgumentMap(this.arguments),
      }),
      type: this.type,
      resolve: (source, _args, _context, info) => source[info.path.key],
    };
  }

  @Memoize()
  public validate(): void {
    if (this.isPublic()) {
      this.getGraphQLFieldConfig();
    }
  }

  protected parseGraphQLArgumentNodes(
    argumentNodes: graphql.FieldNode['arguments'],
    context: GraphQLSelectionContext | undefined,
    path: utils.Path,
  ): Exclude<TArgs, null> {
    if (!this.arguments?.length) {
      if (argumentNodes?.length) {
        throw new utils.GraphError(`Expects no arguments`, { path });
      }

      return undefined as any;
    }

    return utils.parseInputLiterals(
      this.arguments,
      argumentNodes,
      context?.variableValues,
      utils.addPath(path, argsPathKey),
    ) as any;
  }

  public abstract selectGraphQLFieldNode(
    fieldNode: graphql.FieldNode,
    operationContext: OperationContext | undefined,
    selectionContext: GraphQLSelectionContext | undefined,
    path: utils.Path,
  ): SelectionExpression;

  public abstract selectShape(
    value: unknown,
    operationContext: OperationContext | undefined,
    path: utils.Path,
  ): SelectionExpression;
}
