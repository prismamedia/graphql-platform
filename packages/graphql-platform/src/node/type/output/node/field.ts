import type { ComponentOutputType } from './field/component.js';
import type { ReverseEdgeOutputType } from './field/reverse-edge.js';
import type { VirtualOutputType } from './field/virtual.js';

export * from './field/component.js';
export * from './field/reverse-edge.js';
export * from './field/virtual.js';

export type NodeFieldOutputType =
  | ComponentOutputType
  | ReverseEdgeOutputType
  | VirtualOutputType;
