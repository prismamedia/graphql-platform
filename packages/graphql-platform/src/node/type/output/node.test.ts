import { addPath } from '@prismamedia/graphql-platform-utils';
import * as graphql from 'graphql';
import { GraphQLPlatform } from '../../../index.js';
import {
  myAdminContext,
  MyGP,
  myVisitorContext,
  nodeNames,
  nodes,
} from '../../../__tests__/config.js';
import { OperationContext } from '../../operation/context.js';
import { NodeOutputType } from './node.js';

describe(`NodeOutputType`, () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  describe('Definition', () => {
    it.each(nodeNames)('%s has an output type', (nodeName) => {
      const node = gp.getNode(nodeName);

      const nodeOutputType = node.outputType;
      expect(nodeOutputType).toBeInstanceOf(NodeOutputType);

      if (node.isPublic()) {
        expect(nodeOutputType.getGraphQLObjectType()).toBeInstanceOf(
          graphql.GraphQLObjectType,
        );
        expect(
          graphql.printType(nodeOutputType.getGraphQLObjectType()),
        ).toMatchSnapshot();
      } else {
        expect(() => nodeOutputType.getGraphQLObjectType()).toThrowError(
          `The "${nodeName}" node is private`,
        );
      }
    });

    it('throws an Error on duplicate field name', () => {
      expect(
        () =>
          new GraphQLPlatform({
            nodes: {
              ...nodes,
              User: {
                ...nodes.User,
                output: {
                  virtualFields: {
                    username: {
                      args: {},
                      type: new graphql.GraphQLNonNull(graphql.GraphQLString),
                      resolve: () => 'MyUsername',
                    },
                  },
                },
              },
            },
          }),
      ).toThrowError();
    });
  });

  describe('Runtime', () => {
    describe('Fails', () => {
      it.each([
        [
          'Article',
          undefined,
          myAdminContext,
          'Expects an "Article"\'s selection, got: undefined',
        ],
        [
          'Article',
          null,
          myAdminContext,
          'Expects an "Article"\'s selection, got: null',
        ],
        [
          'Article',
          `{ unknownField }`,
          myAdminContext,
          "Expects an \"Article\"'s field, got: 'unknownField'",
        ],
        [
          'Article',
          `{ category { unknownDeepField } }`,
          myAdminContext,
          '"Article.category" - Expects a "Category"\'s field, got: \'unknownDeepField\'',
        ],
        [
          'Article',
          `{ id(first: 5) }`,
          myAdminContext,
          '"Article.id" - Expects no arguments',
        ],
        [
          'Article',
          `{ id { id } }`,
          myAdminContext,
          '"Article.id" - Expects no selection-set',
        ],
        [
          'Article',
          `{ tags { tag { title }} }`,
          myAdminContext,
          '"Article.tags.@args.first" - Expects a non-undefined "UnsignedInt"',
        ],
        [
          'Article',
          `{ category }`,
          myAdminContext,
          '"Article.category" - Expects a "Category"\'s selection',
        ],
        [
          'Article',
          `{ title tags(first: 5) { tag { title } } }`,
          myVisitorContext,
          '"Article.tags" - Unauthorized access to "ArticleTag"',
        ],
      ])(
        '%p.select(%p) throws the error %p',
        (nodeName, input, requestContext, error) => {
          const node = gp.getNode(nodeName);
          const outputType = node.outputType;

          expect(() =>
            outputType.select(
              // @ts-expect-error
              input,
              new OperationContext(gp, requestContext),
              undefined,
              addPath(undefined, node.name),
            ),
          ).toThrowError(error);
        },
      );
    });

    describe('Works', () => {
      it.each<[nodeName: string, fragment: string, selection: string]>([
        [
          'Article',
          `{ id }`,
          `{
  id
}`,
        ],
        [
          'Article',
          `... { id }`,
          `{
  id
}`,
        ],
        [
          'Article',
          `... on Article { id }`,
          `{
  id
}`,
        ],
        [
          'Article',
          `fragment MyTestFragment on Article { id }`,
          `{
  id
}`,
        ],
        [
          'Article',
          `{ id lowerCasedTitle category { id _a: parent { _id } } }`,
          `{
  id
  status
  title
  category {
    title
    id
    parent {
      _id
    }
  }
}`,
        ],
        [
          'User',
          `{
            username
            profile {
              birthday
              twitterHandle
            }
            ... {
              profile {
                facebookId
                birthday
              }
            }
            createdArticles(where: { status: published }, orderBy: [updatedAt_DESC], first: 10) {
              _id
              id
            }
            createdArticleCount(where: { status: published })
          }`,
          `{
  username
  profile {
    birthday
    twitterHandle
    facebookId
  }
  createdArticles(first: 10) {
    _id
    id
  }
  createdArticleCount
}`,
        ],
      ])('%p.select(%p) = %p', (nodeName, fragment, selectionSet) => {
        const node = gp.getNode(nodeName);
        const outputType = node.outputType;

        expect(
          graphql.print(outputType.select(fragment).toGraphQLSelectionSet()),
        ).toEqual(selectionSet);
      });
    });
  });
});
