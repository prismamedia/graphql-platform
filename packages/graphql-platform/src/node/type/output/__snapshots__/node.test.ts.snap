// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeOutputType Definition Article has an output type 1`] = `
""""The article is the main resource, written by the journalists"""
type Article implements PublicNodeInterface {
  """This UUID identifies an Article publicly"""
  id: UUIDv4!
  status: ArticleStatus!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!

  """The article's body"""
  body: DraftJS
  category: Category
  createdBy: User
  createdAt: DateTime!
  updatedBy: User
  updatedAt: DateTime!

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  tags(where: ArticleTagFilterInput, orderBy: [ArticleTagOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [ArticleTag!]!

  """
  Number of "tags"
  """
  tagCount(where: ArticleTagFilterInput): UnsignedInt!

  """A custom field with a dependency"""
  lowerCasedTitle: NonEmptyTrimmedString!

  """A custom field with a dependency"""
  upperCasedTitle: NonEmptyTrimmedString!
}"
`;

exports[`NodeOutputType Definition ArticleTag has an output type 1`] = `
"type ArticleTag {
  article: Article!
  tag: Tag!
  order: UnsignedInt!
  moderations(where: ArticleTagModerationFilterInput, skip: UnsignedInt, first: UnsignedInt!): [ArticleTagModeration!]!

  """
  Number of "moderations"
  """
  moderationCount(where: ArticleTagModerationFilterInput): UnsignedInt!
}"
`;

exports[`NodeOutputType Definition ArticleTagModeration has an output type 1`] = `
"type ArticleTagModeration {
  articleTag: ArticleTag!
  moderator: User!
  moderation: NonEmptyTrimmedString
}"
`;

exports[`NodeOutputType Definition Category has an output type 1`] = `
"type Category implements PublicNodeInterface {
  id: UUIDv4!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
  parent: Category
  order: UnsignedInt!

  """This category's children"""
  children(where: CategoryFilterInput, orderBy: [CategoryOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [Category!]!

  """
  Number of "children"
  """
  childCount(where: CategoryFilterInput): UnsignedInt!

  """The articles attached to this category"""
  articles(where: ArticleFilterInput, orderBy: [ArticleOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [Article!]!

  """
  Number of "articles"
  """
  articleCount(where: ArticleFilterInput): UnsignedInt!
}"
`;

exports[`NodeOutputType Definition Tag has an output type 1`] = `
"type Tag implements PublicNodeInterface {
  id: UUIDv4!

  """A tag can be deprecated"""
  deprecated: Boolean
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
  createdAt: DateTime!
  updatedAt: DateTime!
  articles(where: ArticleTagFilterInput, orderBy: [ArticleTagOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [ArticleTag!]!

  """
  Number of "articles"
  """
  articleCount(where: ArticleTagFilterInput): UnsignedInt!
}"
`;

exports[`NodeOutputType Definition User has an output type 1`] = `
"type User implements PublicNodeInterface {
  id: UUIDv4!
  username: NonEmptyTrimmedString!

  """All the articles this user has created"""
  createdArticles(where: ArticleFilterInput, orderBy: [ArticleOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [Article!]!

  """
  Number of "createdArticles"
  """
  createdArticleCount(where: ArticleFilterInput): UnsignedInt!

  """All the articles this user has updated"""
  updatedArticles(where: ArticleFilterInput, orderBy: [ArticleOrderingInput!], skip: UnsignedInt, first: UnsignedInt!): [Article!]!

  """
  Number of "updatedArticles"
  """
  updatedArticleCount(where: ArticleFilterInput): UnsignedInt!

  """This user's profile, only the optional informations"""
  profile: UserProfile
  articleTagModerations(where: ArticleTagModerationFilterInput, skip: UnsignedInt, first: UnsignedInt!): [ArticleTagModeration!]!

  """
  Number of "articleTagModerations"
  """
  articleTagModerationCount(where: ArticleTagModerationFilterInput): UnsignedInt!
}"
`;

exports[`NodeOutputType Definition UserProfile has an output type 1`] = `
"type UserProfile {
  user: User!
  birthday: Date
  facebookId: ID
  googleId: ID
  twitterHandle: ID
}"
`;
